src-self-hosted/link.zig

```diff
@@ -603,24 +603,62 @@ pub const File = struct {
             return min_pos - start;
         }
 
-        fn findFreeSpace(self: *Elf, object_size: u64, min_alignment: u16) u64 {
+        fn inRange(self: Elf, start: u64) ?Options.AddressSpaceEntry {
+            if (self.base.options.address_space) |address_space| {
+                for (address_space.items) |a| {
+                    // Start is inclusive, end is exclusive
+                    if (start >= a.start and start < a.end) {
+                        return a;
+                    }
+                }
+            }
+            return null;
+        }
+
+        fn rangeAfter(self: Elf, addr: u64) ?u64 {
+            if (self.base.options.address_space) |address_space| {
+                for (address_space.items) |a| {
+                    if (addr < a.start) {
+                        return a.start;
+                    }
+                }
+            }
+            return null;
+        }
+
+        fn findFreeSpace(self: *Elf, object_size: u64, min_alignment: u16) ?u64 {
```

Comment:
This function finds free space in the ELF file, not virtual address space.

The allocation functions for virtual address are related to `TextBlock`.